packages/core/src/lib/filter/v2/index.ts

```diff
@@ -0,0 +1,334 @@
+ * - https://github.com/status-im/go-waku/issues/245
+ * - https://github.com/status-im/nwaku/issues/948
+ */
+class FilterV2 extends BaseProtocol implements IFilterV2 {
+  options: ProtocolCreateOptions;
+  private subscriptions: Map<RequestID, unknown>;
+
+  constructor(public libp2p: Libp2p, options?: ProtocolCreateOptions) {
+    super(
+      FilterV2Codecs.SUBSCRIBE,
+      libp2p.peerStore,
+      libp2p.getConnections.bind(libp2p)
+    );
+
+    this.libp2p.handle(FilterV2Codecs.PUSH, this.onRequest.bind(this));
+
+    this.options = options ?? {};
+    this.subscriptions = new Map();
+  }
+
+  /**
+   * @param decoders Decoder or array of Decoders to use to decode messages, it also specifies the content topics.
+   * @param callback A function that will be called on each message returned by the filter.
+   * @param opts The FilterSubscriptionOpts used to narrow which messages are returned, and which peer to connect to.
+   * @returns Unsubscribe function that can be used to end the subscription.
+   */
+  async subscribe<T extends IDecodedMessage>(
+    decoders: IDecoder<T> | IDecoder<T>[],
+    callback: Callback<T>,
+    opts?: ProtocolOptions
+  ): Promise<UnsubscribeFunction> {
+    const decodersArray = Array.isArray(decoders) ? decoders : [decoders];
+    const { pubSubTopic = DefaultPubSubTopic } = this.options;
+
+    const contentTopics = Array.from(groupByContentTopic(decodersArray).keys());
+
+    const request = FilterSubscribeRpc.createSubscribeRequest(
+      pubSubTopic,
+      contentTopics
+    );
+
+    const { requestId } = request;
+
+    const peer = await this.getPeer(opts?.peerId);
+    const stream = await this.newStream(peer);
+
+    try {
+      const res = await pipe(
+        [request.encode()],
+        lp.encode(),
+        stream,
+        lp.decode(),
+        async (source) => await all(source)
+      );
+
+      const { statusCode, requestId } = FilterSubscribeResponse.decode(
+        res[0].slice()
+      );
+
+      if (statusCode < 200 || statusCode >= 300) {
+        throw new Error(
+          `Filter subscribe request ${requestId} failed with status code ${statusCode}`
+        );
+      }
+
+      log(
+        "Subscribed to peer ",
+        peer.id.toString(),
+        "for content topics",
+        contentTopics
+      );
+    } catch (e) {
+      log(
+        "Error subscribing to peer ",
+        peer.id.toString(),
+        "for content topics",
+        contentTopics,
+        ": ",
+        e
+      );
+      throw e;
+    }
+
+    const subscription: Subscription<T> = {
+      callback,
+      decoders: decodersArray,
+      pubSubTopic,
+    };
+    this.subscriptions.set(requestId, subscription);
+
+    return async () => {
+      await this.unsubscribe(pubSubTopic, contentTopics, requestId, peer);
+      this.subscriptions.delete(requestId);
+    };
+  }
+
+  public async unsubscribeAll(): Promise<void> {
```

Comment:
I think a peer Id should be part of the API. Otherwise, we are just unsubscribing to a random peer, which makes no sense.